File: app/src/main/java/com/example/project01_group05/api/MangaRepository.java
package com.example.project01_group05.api;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MangaRepository {

    public void fetchManga(MangaCallback callback) {

        MangaDexApi api = MangaDexClient.getApi();

        api.getManga(10).enqueue(new Callback<MangaResponse>() {

            @Override
            public void onResponse(
                    Call<MangaResponse> call,
                    Response<MangaResponse> response) {

                if (response.isSuccessful() && response.body() != null) {

                    List<MangaData> mangaList = response.body().getData();

                    callback.onSuccess(mangaList);

                } else {

                    callback.onError(
                            "MangaDex request failed. HTTP code: "
                                    + response.code()
                    );
                }
            }

            @Override
            public void onFailure(
                    Call<MangaResponse> call,
                    Throwable throwable) {

                String message = throwable.getMessage();

                if (message == null) {
                    message = "Unable to retrieve manga from MangaDex.";
                }

                callback.onError(message);
            }
        });
    }
}

File: app/src/main/java/com/example/project01_group05/api/MangaRepository.kt
package com.example.project01_group05.api

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MangaRepository : MangaSearchRepository{

    fun fetchManga(callback: MangaCallback) {

        val api = MangaDexClient.getApi()

        api.getManga(10).enqueue(object : Callback<MangaResponse> {

            override fun onResponse(
                call: Call<MangaResponse>,
                response: Response<MangaResponse>
            ) {
                if (response.isSuccessful && response.body() != null) {

                    val mangaList = response.body()!!.data
                    callback.onSuccess(mangaList)

                } else {

                    callback.onError(
                        "MangaDex request failed. HTTP code: ${response.code()}"
                    )
                }
            }

            override fun onFailure(
                call: Call<MangaResponse>,
                throwable: Throwable
            ) {
                val message =
                    throwable.message ?: "Unable to retrieve manga from MangaDex."

                callback.onError(message)
            }
        })
    }

    override fun searchManga(
        title: String,
        callback: MangaCallback
    ) {
        val api = MangaDexClient.getApi()

        api.searchManga(title).enqueue(object : Callback<MangaResponse> {

            override fun onResponse(
                call: Call<MangaResponse>,
                response: Response<MangaResponse>
            ) {
                if (response.isSuccessful && response.body() != null) {

                    val mangaList = response.body()!!.data
                    callback.onSuccess(mangaList)

                } else {

                    callback.onError(
                        "MangaDex search failed. HTTP code: ${response.code()}"
                    )
                }
            }

            override fun onFailure(
                call: Call<MangaResponse>,
                throwable: Throwable
            ) {
                val message =
                    throwable.message ?: "Unable to search MangaDex."

                callback.onError(message)
            }
        })
    }
}